src/lib/electricity-maps/index.ts

```diff
@@ -0,0 +1,119 @@
+                const blocs = [...Array(hours).keys()]
+                // Calculate for each full hour the ratio of the hour that is in the bloc
+                const hourly_ratios = blocs.map((hour) => {
+                    // this is the first hourly bloc, the ratio is the time between the start and the next hour.
+                    if (hour === 0) {
+                        return start.add(1, 'hour').startOf('hour').diff(start, 'second') / 3600;
+                    }
+                    // this is the last hourly bloc, the ratio is the time between the start of the hour and the end.
+                    if (hour === hours - 1) {
+                        return end.diff(start.add(hour, 'hour').startOf('hour'), 'second') / 3600;
+                    }
+                    return 1;
+                });
+                let total_carbon_intensity = 0;
+                carbon_intensities.then((test) =>
+                    test.forEach(
+                    (carbon_intensity, index)=>{
+                        total_carbon_intensity += carbon_intensity.value * hourly_ratios[index];
+                    }
+                ));
+                return {
+                    ...model_param,
+                    carbon_intensity: total_carbon_intensity * power_consumption,
+                    unit: unit
+                }
+            });
+    }
+    initAuth() {
+        dotenv.config();
+        const token = process.env.EMAPS_TOKEN;
+        if (!token) {
+            throw new AuthorizationError(
+                this.errorBuilder({
+                    message: 'Missing token',
+                    scope: 'authorization',
+                })
+            );
+        }
+        this.authorizationHeader = `auth-token: ${token}`;
+    }
+    private async get_carbon_intensity(longitude: number, latitude: number, start: dayjs.Dayjs, end: dayjs.Dayjs): Promise<KeyValuePair[]> {
+        const parameters = {
+            lon: longitude,
+            lat: latitude,
+            start: start,
+            end: end
+        }
+        if (end.diff(start, 'days') > 10) {
+            throw InputValidationError(
+                this.errorBuilder({
+                    message: 'The maximum duration is 10 days for the Emaps API.',
+                    scope: 'input',
+                })
+            )
+        }
+        const url = `${BASE_URL}/carbon-intensity/past-range`;
+        return axios.get(
+            url,
+            {
+                params: parameters,
+                headers: {
+                    'auth-token': this.authorizationHeader,
+                }
+            }
+        ).then(
+            (response)=>{
+                if (response.status !== 200) {
+                    throw new APIRequestError(
+                        this.errorBuilder({
+                            message: `Error: ${JSON.stringify(response.status)}`,
+                        })
+                    );
+                }
```

Comment:
Do we want to handle more response codes here explicitly?
Like can we use the `AuthorizationError` to throw a Autorization error if the token is invalid and we return a 401 status for example.

Reply:
Yes we could add this but rather in a subsequent PR.